### PR TITLE
Chart enhancement

### LINKED ISSUE #1072: Enhance helm chart to support more scenes
**What would you like to be added**:
1. support to install karmada host into any namespace of cluster
I only have one host cluster for testing Karmada, but I want to test different configurations. So I would like to install it into different namespaces.

2. support other service type of Karmada-apiserver, such as NodePort, LoadBalancer

3. ETCD might need multiple replicaCount, and PVC to storage data
**Why is this needed**:

### PR DESCRIPTION
**Pull request by JackZxj**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Enhance helm chart to support more scenes

**Which issue(s) this PR fixes**:
Fixes #1072

**Special notes for your reviewer**:
By default, it is no different from the previous version. After adding some parameters in chart, we can support the deployment of multiple on the same cluster through different namespaces (for testing or...), support different karmada-apiserver service types, and support hostpath or pvc for ETCD, support multiple ETCD replicas.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by jrkeen** [comment]
Thanks for your commits, because CI does not include Helm test, put the test results as much as possible, such as: #779 (comment)

**Comment by JackZxj** [comment]
> 

 👌 I got it!

**Comment by JackZxj** [comment]
I did some tests in kind. Are these enough? @jrkeen 

```bash
[root@10 karmada]# export KUBECONFIG=/root/karmada/pr_test/pr.config
[root@10 karmada]# kubectl get no -owide
NAME               STATUS   ROLES                  AGE   VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE       KERNEL-VERSION          CONTAINER-RUNTIME
pr-control-plane   Ready    control-plane,master   51m   v1.20.2   172.19.0.3    <none>        Ubuntu 20.10   3.10.0-957.el7.x86_64   containerd://1.4.0-106-gce4439a8
pr-worker          Ready    <none>                 51m   v1.20.2   172.19.0.6    <none>        Ubuntu 20.10   3.10.0-957.el7.x86_64   containerd://1.4.0-106-gce4439a8
pr-worker2         Ready    <none>                 51m   v1.20.2   172.19.0.7    <none>        Ubuntu 20.10   3.10.0-957.el7.x86_64   containerd://1.4.0-106-gce4439a8

###################################################################
# 1. all by default, install into karmada-system just like before
###################################################################
[root@10 karmada]# helm install karmada -n karmada-system ./charts/ --create-namespace
NAME: karmada
LAST DEPLOYED: Mon Dec  6 16:51:40 2021
NAMESPACE: karmada-system
STATUS: deployed
REVISION: 1
TEST SUITE: None
[root@10 karmada]# kubectl  get po -n karmada-system -owide
NAME                                               READY   STATUS    RESTARTS   AGE   IP           NODE         NOMINATED NODE   READINESS GATES
etcd-0                                             1/1     Running   0          90s   10.244.1.3   pr-worker    <none>           <none>
karmada-apiserver-58bd66c89-x5rkk                  1/1     Running   0          90s   172.19.0.7   pr-worker2   <none>           <none>
karmada-controller-manager-7dbd7f59b-778wl         1/1     Running   3          90s   10.244.1.6   pr-worker    <none>           <none>
karmada-kube-controller-manager-78cf959d84-6xlrw   1/1     Running   3          90s   10.244.2.4   pr-worker2   <none>           <none>
karmada-scheduler-56c7d786bc-dqmwj                 1/1     Running   0          90s   10.244.2.3   pr-worker2   <none>           <none>
karmada-webhook-7f4b95875d-cdtsg                   1/1     Running   3          90s   10.244.1.4   pr-worker    <none>           <none>

[root@10 karmada]# kubectl get secret -n karmada-system karmada-kubeconfig -o jsonpath={.data.kubeconfig} | base64 -d > default.config
[root@10 karmada]# vi default.config
# remove 'certificate-authority-data', make 'insecure-skip-tls-verify' to be true, change 'server' to 172.19.0.7:5443
[root@10 karmada]# kubectl get crd --kubeconfig default.config
NAME                                                         CREATED AT
clusteroverridepolicies.policy.karmada.io                    2021-12-06T08:52:54Z
clusterpropagationpolicies.policy.karmada.io                 2021-12-06T08:52:54Z
clusterresourcebindings.work.karmada.io                      2021-12-06T08:52:54Z
clusters.cluster.karmada.io                                  2021-12-06T08:52:54Z
overridepolicies.policy.karmada.io                           2021-12-06T08:52:54Z
propagationpolicies.policy.karmada.io                        2021-12-06T08:52:54Z
replicaschedulingpolicies.policy.karmada.io                  2021-12-06T08:52:54Z
resourcebindings.work.karmada.io                             2021-12-06T08:52:54Z
resourceinterpreterwebhookconfigurations.config.karmada.io   2021-12-06T08:52:54Z
serviceexports.multicluster.x-k8s.io                         2021-12-06T08:52:54Z
serviceimports.multicluster.x-k8s.io                         2021-12-06T08:52:54Z
works.work.karmada.io                                        2021-12-06T08:52:54Z


###################################################################
# 2. deploy by NodePort, in alpha namespace
###################################################################
[root@10 karmada]# helm install karmada-alpha -n alpha ./charts/ --create-namespace --set apiServer.hostNetwork=false --set apiServer.serviceType=NodePort
NAME: karmada-alpha
LAST DEPLOYED: Mon Dec  6 17:19:15 2021
NAMESPACE: alpha
STATUS: deployed
REVISION: 1
TEST SUITE: None
[root@10 pr_test]# kubectl get po,svc -owide -n alpha
NAME                                                        READY   STATUS    RESTARTS   AGE   IP            NODE         NOMINATED NODE   READINESS GATES
pod/etcd-0                                                  1/1     Running   0          63s   10.244.2.8    pr-worker2   <none>           <none>
pod/karmada-alpha-apiserver-849777f585-j6t7b                1/1     Running   0          63s   10.244.1.11   pr-worker    <none>           <none>
pod/karmada-alpha-controller-manager-849d74bd96-52dbq       1/1     Running   2          63s   10.244.2.10   pr-worker2   <none>           <none>
pod/karmada-alpha-kube-controller-manager-b79c4d568-8hn8l   1/1     Running   2          63s   10.244.1.10   pr-worker    <none>           <none>
pod/karmada-alpha-scheduler-c6bc4f9bf-2qcrl                 1/1     Running   0          63s   10.244.1.9    pr-worker    <none>           <none>
pod/karmada-alpha-webhook-6fcf69945-4smm7                   1/1     Running   2          63s   10.244.2.9    pr-worker2   <none>           <none>

NAME                              TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)             AGE   SELECTOR
service/etcd                      ClusterIP   None            <none>        2379/TCP,2380/TCP   63s   app=etcd
service/etcd-client               ClusterIP   10.96.98.89     <none>        2379/TCP            63s   app=etcd
service/karmada-alpha-apiserver   NodePort    10.96.197.120   <none>        5443:31154/TCP      63s   app=karmada-apiserver
service/karmada-alpha-webhook     ClusterIP   10.96.76.19     <none>        443/TCP             63s   app=karmada-alpha-webhook

[root@10 karmada]# kubectl get secret -n alpha karmada-alpha-kubeconfig -o jsonpath={.data.kubeconfig} | base64 -d > nodeport.config
[root@10 karmada]# vi nodeport.config
# remove 'certificate-authority-data', make 'insecure-skip-tls-verify' to be true, change 'server' to 172.19.0.3:31154
[root@10 karmada]# kubectl get crd --kubeconfig nodeport.config
NAME                                                         CREATED AT
clusteroverridepolicies.policy.karmada.io                    2021-12-06T09:20:15Z
clusterpropagationpolicies.policy.karmada.io                 2021-12-06T09:20:15Z
clusterresourcebindings.work.karmada.io                      2021-12-06T09:20:16Z
clusters.cluster.karmada.io                                  2021-12-06T09:20:17Z
overridepolicies.policy.karmada.io                           2021-12-06T09:20:18Z
propagationpolicies.policy.karmada.io                        2021-12-06T09:20:19Z
replicaschedulingpolicies.policy.karmada.io                  2021-12-06T09:20:19Z
resourcebindings.work.karmada.io                             2021-12-06T09:20:20Z
resourceinterpreterwebhookconfigurations.config.karmada.io   2021-12-06T09:20:21Z
serviceexports.multicluster.x-k8s.io                         2021-12-06T09:20:21Z
serviceimports.multicluster.x-k8s.io                         2021-12-06T09:20:21Z
works.work.karmada.io                                        2021-12-06T09:20:22Z


###################################################################
# 3. deploy by LoadBalancer, multiple etcd with pvc, in mybeta namespace
###################################################################
[root@10 pr_test]# kubectl apply -f pv0.yaml
persistentvolume/pv-0 created
[root@10 pr_test]# kubectl apply -f pv1.yaml
persistentvolume/pv-1 created
[root@10 pr_test]# kubectl apply -f pv2.yaml
persistentvolume/pv-2 created
[root@10 karmada]# kubectl get pv
NAME   CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS      CLAIM   STORAGECLASS   REASON   AGE
pv-0   2Gi        RWO            Retain           Available           standard                55s
pv-1   2Gi        RWO            Retain           Available           standard                48s
pv-2   2Gi        RWO            Retain           Available           standard                42s

[root@10 karmada]# helm install hello-beta -n mybeta ./charts/ --create-namespace --set apiServer.hostNetwork=false \
--set apiServer.serviceType=LoadBalancer \
--set etcd.internal.replicaCount=3 \
--set etcd.internal.storageType=pvc \
--set etcd.internal.pvc.storageClass=standard \
--set etcd.internal.pvc.size=2Gi
NAME: hello-beta
LAST DEPLOYED: Mon Dec  6 21:20:36 2021
NAMESPACE: mybeta
STATUS: deployed
REVISION: 1
TEST SUITE: None

[root@10 karmada]# kubectl get po,svc,pvc -n mybeta
NAME                                                      READY   STATUS    RESTARTS   AGE
pod/etcd-0                                                1/1     Running   0          35m
pod/etcd-1                                                1/1     Running   0          35m
pod/etcd-2                                                1/1     Running   0          35m
pod/hello-beta-apiserver-7fbcf69cbd-p92qb                 1/1     Running   0          35m
pod/hello-beta-controller-manager-7bfc5bcdc8-kqbfz        1/1     Running   3          35m
pod/hello-beta-kube-controller-manager-6dc66587fd-krs6d   1/1     Running   3          35m
pod/hello-beta-scheduler-56d6c765bc-k9jk4                 1/1     Running   0          35m
pod/hello-beta-webhook-5df9484654-95zbc                   1/1     Running   3          35m

NAME                           TYPE           CLUSTER-IP      EXTERNAL-IP      PORT(S)             AGE
service/etcd                   ClusterIP      None            <none>           2379/TCP,2380/TCP   35m
service/etcd-client            ClusterIP      10.96.61.204    <none>           2379/TCP            35m
service/hello-beta-apiserver   LoadBalancer   10.96.210.1     172.19.252.100   5443:32725/TCP      35m
service/hello-beta-webhook     ClusterIP      10.96.186.225   <none>           443/TCP             35m

NAME                                     STATUS   VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS   AGE
persistentvolumeclaim/etcd-data-etcd-0   Bound    pv-0     2Gi        RWO            standard       35m
persistentvolumeclaim/etcd-data-etcd-1   Bound    pv-2     2Gi        RWO            standard       35m
persistentvolumeclaim/etcd-data-etcd-2   Bound    pv-1     2Gi        RWO            standard       35m

[root@10 karmada]# kubectl get secret -n mybeta hello-beta-kubeconfig -o jsonpath={.data.kubeconfig} | base64 -d > lb.config
[root@10 karmada]# vi lb.config
# remove 'certificate-authority-data', make 'insecure-skip-tls-verify' to be true, change 'server' to 172.19.252.100:5443
[root@10 karmada]# kubectl get crd --kubeconfig lb.config
NAME                                                         CREATED AT
clusteroverridepolicies.policy.karmada.io                    2021-12-06T13:21:23Z
clusterpropagationpolicies.policy.karmada.io                 2021-12-06T13:21:23Z
clusterresourcebindings.work.karmada.io                      2021-12-06T13:21:23Z
clusters.cluster.karmada.io                                  2021-12-06T13:21:24Z
overridepolicies.policy.karmada.io                           2021-12-06T13:21:24Z
propagationpolicies.policy.karmada.io                        2021-12-06T13:21:25Z
replicaschedulingpolicies.policy.karmada.io                  2021-12-06T13:21:25Z
resourcebindings.work.karmada.io                             2021-12-06T13:21:25Z
resourceinterpreterwebhookconfigurations.config.karmada.io   2021-12-06T13:21:26Z
serviceexports.multicluster.x-k8s.io                         2021-12-06T13:21:26Z
serviceimports.multicluster.x-k8s.io                         2021-12-06T13:21:26Z
works.work.karmada.io                                        2021-12-06T13:21:27Z

###################################################################
# install three karmada control planes in one cluster
###################################################################
[root@10 karmada]# helm ls -A
NAME            NAMESPACE       REVISION        UPDATED                                 STATUS          CHART           APP VERSION
hello-beta      mybeta          1               2021-12-06 21:20:36.08825958 +0800 CST  deployed        karmada-0.0.2   v0.10.1
karmada         karmada-system  1               2021-12-06 16:51:40.110854362 +0800 CST deployed        karmada-0.0.2   v0.10.1
karmada-alpha   alpha           1               2021-12-06 17:19:15.311936404 +0800 CST deployed        karmada-0.0.2   v0.10.1
```

File: charts/values.yaml [hunk 1]
```diff
@@ -216,6 +220,11 @@ apiServer:
   tolerations: [ ]
     # - key: node-role.kubernetes.io/master
     #   operator: Exists
+  ## @param apiServer.serviceType default service type for apiserver
+  ## "LoadBalancer" means using LoadBalancer
+  ## "ClusterIP" means using ClusterIP
+  ## "NodePort" means using NodePort
+  serviceType: ClusterIP
```
Comment (jrkeen) [inline charts/values.yaml:227]: Is it possible to custom the port of the nodeport?

Comment (JackZxj) [reply charts/values.yaml:227]: I think that `NodePort` just for testing in development, so I let it be allocated automatically. If you think it is necessary, I can add this.

Comment (jrkeen) [reply charts/values.yaml:227]: Configurable will support more scenes, automatically assign if it is not set. :)

Comment (JackZxj) [reply charts/values.yaml:227]: I have added it.

**Comment by jrkeen** [comment]
Thanks very much, you should squash your commits

**Comment by JackZxj** [comment]
> Thanks very much, you should squash your commits

Sorry that I thought you would squash when this PR be merged. Is it okay now?

**Comment by jrkeen** [comment]
/lgtm
cc @algebra2k

**Comment by RainbowMango** [comment]
Thanks and welcome @JackZxj

**Comment by JackZxj** [comment]
这是来自QQ邮箱的自动回复邮件。您好，您的邮件我已经收到，我会尽快做出处理的。by Zhang XJ.

**Comment by JackZxj** [comment]
> Thanks and welcome @JackZxj

I'm glad to make it better~

**Pull request merged**